src/full/Agda/TypeChecking/DiscrimTree.hs

```diff
@@ -0,0 +1,359 @@
+import Agda.TypeChecking.DiscrimTree.Types
+
+import qualified Agda.Utils.ProfileOptions as Profile
+
+import Agda.Utils.Impossible
+import Agda.Utils.Trie (Trie(..))
+
+-- | Dummy term to use as a stand-in for expanded eta-records while
+-- building instance trees.
+etaExpansionDummy :: Term
+etaExpansionDummy = Dummy "eta-record argument in instance head" []
+
+-- | Extract a list of arguments from the list of eliminations; If
+-- called while *adding* an instance, additionally replace any arguments
+-- that might belong to an eta-record by dummy terms.
+termKeyElims
+  :: Bool     -- ^ Are we adding or looking up an instance?
+  -> TCM Type -- ^ Continuation to compute the type of the arguments in the spine.
+  -> Elims    -- ^ The spine.
+  -> MaybeT TCM (Int, [Term])
+
+-- Since the case tree was generated with wildcards everywhere an eta
+-- record appeared, if we're *looking up* an instance, we don't have to
+-- do the censorship again.
+termKeyElims False _ elims = do
+  es <- MaybeT $ pure (allApplyElims elims)
+  pure (length es, map unArg es)
+
+termKeyElims precise ty elims = do
+  args <- MaybeT $ pure (allApplyElims elims)
+
+  let
+    go ty (Arg _ a:as) = flip (ifPiTypeB ty) (const mzero) \dom ty' -> do
+
+      -- Is this argument an eta record type --- or a blocked value? In
+      -- either case, we replace this position by a dummy, to make sure
+      -- that eta-equality is respected.
+      maybeEta <- ifBlocked (unDom dom) (\_ _ -> pure True) \_ tm ->
+        isJust <$> isEtaRecordType tm
+
+      let
+        here
+          | maybeEta  = etaExpansionDummy
+          | otherwise = a
+
+      (k, there) <- addContext dom (go (unAbs ty') as)
+      pure (k + 1, here:there)
+
+    go _ [] = pure (0, [])
+
+  ty <- lift ty
+  go ty args
+
+-- | Ticky profiling for the reason behind "inexactness" in instance
+-- search. If at some point while narrowing the set of candidates we had
+-- to go through all the possibilities, one of these counters is
+-- incremented.
+tickExplore :: Term -> TCM ()
+tickExplore tm = whenProfile Profile.Instances do
+  tick "flex term blocking instance"
+
+  case tm of
+    Def{}      -> tick "explore: Def"
+    Var{}      -> tick "explore: Var"
+    Lam _ v
+      -- These two are a hunch: just like FunK, it might be worth
+      -- optimising for the case where a lambda is constant (which is
+      -- easy to handle, by just pretending the term is something else).
+      -- These would come up in e.g. Dec (PathP (λ i → Nat) x y)
+      | NoAbs{} <- v -> tick "explore: constant function"
+      | Abs _ b <- v, not (0 `freeIn` b) -> tick "explore: constant function"
+
+      | otherwise    -> tick "explore: Lam"
+    Lit{}      -> tick "explore: Lit"
+    Sort{}     -> tick "explore: Sort"
+    Level{}    -> tick "explore: Level"
+    MetaV{}    -> tick "explore: Meta"
+    DontCare{} -> tick "explore: DontCare"
+    _ -> pure ()
+
+-- | Checks whether the given abstraction could have been 'NoAbs'.
```

Comment:
Well, my `Maybe` version has the benefit of doing the right `strengthen` dance to get the body out, though this incurs an extra `Subst a` constraint over `isBinderUsed`. I could merge these two (`isBinderUsed = isNothing . isNoAbs`), since `isBinderUsed` is only used concretely at `Term` and polymorphically in `compareDom`, which could survive the extra `Subst` constraint.